tests/test_launcher.py

```diff
@@ -0,0 +1,147 @@
+"""This module contains tests for aea launcher."""
+
+import os
+import shutil
+import tempfile
+import time
+from pathlib import Path
+
+import pytest
+
+import yaml
+
+from aea.cli.core import cli
+from aea.configurations.base import DEFAULT_AEA_CONFIG_FILE
+from aea.helpers.base import cd
+from aea.launcher import AEALauncher
+from aea.test_tools.click_testing import CliRunner
+from aea.test_tools.test_cases import CLI_LOG_OPTION
+
+
+from tests.common.utils import wait_for_condition
+from tests.conftest import AUTHOR, CUR_PATH
+
+
+class TestThreadLauncherMode:
+    """Test launcher in threaded mode."""
+
+    RUNNER_MODE = "thread"
+
+    @classmethod
+    def setup_class(cls):
+        """Set the test up."""
+        cls.runner = CliRunner()
+        cls.agent_name_1 = "myagent_1"
+        cls.agent_name_2 = "myagent_2"
+        cls.failing_agent = "failing_agent"
+        cls.cwd = os.getcwd()
+        cls.t = tempfile.mkdtemp()
+        os.chdir(cls.t)
+
+        result = cls.runner.invoke(
+            cli, [*CLI_LOG_OPTION, "init", "--local", "--author", AUTHOR]
+        )
+        assert result.exit_code == 0
+
+        result = cls.runner.invoke(
+            cli, [*CLI_LOG_OPTION, "create", "--local", cls.agent_name_1]
+        )
+        assert result.exit_code == 0
+
+        result = cls.runner.invoke(
+            cli, [*CLI_LOG_OPTION, "create", "--local", cls.agent_name_2]
+        )
+        assert result.exit_code == 0
+
+        result = cls.runner.invoke(
+            cli, [*CLI_LOG_OPTION, "create", "--local", cls.failing_agent]
+        )
+        assert result.exit_code == 0
+
+        os.chdir(cls.failing_agent)
+        shutil.copytree(
+            Path(CUR_PATH, "data", "exception_skill"),
+            Path(cls.t, cls.failing_agent, "skills", "exception"),
+        )
+        config_path = Path(cls.t, cls.failing_agent, DEFAULT_AEA_CONFIG_FILE)
+        config = yaml.safe_load(open(config_path))
+        config.setdefault("skills", []).append("fetchai/exception:0.1.0")
+        yaml.safe_dump(config, open(config_path, "w"))
+        os.chdir(cls.t)
+
+        for agent_name in (cls.agent_name_1, cls.agent_name_2, cls.failing_agent):
+            cls.set_runtime_mode_to_async(agent_name)
+
+    @classmethod
+    def set_runtime_mode_to_async(cls, agent_name: str) -> None:
+        """Set runtime mode of the agent to async."""
+        with cd(agent_name):
+            config_path = Path(cls.t, agent_name, DEFAULT_AEA_CONFIG_FILE)  # type: ignore # mypy does not see setup_class
+            config = yaml.safe_load(open(config_path))
+            config.setdefault("runtime_mode", "async")
+            yaml.safe_dump(config, open(config_path, "w"))
+
+    @classmethod
+    def teardown_class(cls):
+        """Tear the test down."""
+        os.chdir(cls.cwd)
+        try:
+            shutil.rmtree(cls.t)  # type: ignore # mypy does not see setup_class
```

Comment:
You can define t as class variables on the class and it should work..same for the others

Reply:
not sure it's required, it's a test code and mypy does not follow pytest workflow, this causes this missing definition of some values. but will check